In [1]:
import pandas as pd

df = pd.read_csv('/content/car data.csv')

df.head()

,Car_Name,Year,Selling_Price,Present_Price,Kms_Driven,Fuel_Type,Seller_Type,Transmission,Owner
0,ritz,2014,3.35,5.59,27000,Petrol,Dealer,Manual,0
1,sx4,2013,4.75,9.54,43000,Diesel,Dealer,Manual,0
2,ciaz,2017,7.25,9.85,6900,Petrol,Dealer,Manual,0
3,wagon r,2011,2.85,4.15,5200,Petrol,Dealer,Manual,0
4,swift,2014,4.60,6.87,42450,Diesel,Dealer,Manual,0


In [2]:
df.shape

(301, 9)

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 301 entries, 0 to 300
Data columns (total 9 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Car_Name       301 non-null    object 
 1   Year           301 non-null    int64  
 2   Selling_Price  301 non-null    float64
 3   Present_Price  301 non-null    float64
 4   Kms_Driven     301 non-null    int64  
 5   Fuel_Type      301 non-null    object 
 6   Seller_Type    301 non-null    object 
 7   Transmission   301 non-null    object 
 8   Owner          301 non-null    int64  
dtypes: float64(2), int64(3), object(4)
memory usage: 21.3+ KB


In [4]:
df.columns

Index(['Car_Name', 'Year', 'Selling_Price', 'Present_Price', 'Kms_Driven',
       'Fuel_Type', 'Seller_Type', 'Transmission', 'Owner'],
      dtype='object')

In [5]:
df.tail()

,Car_Name,Year,Selling_Price,Present_Price,Kms_Driven,Fuel_Type,Seller_Type,Transmission,Owner
296,city,2016,9.50,11.6,33988,Diesel,Dealer,Manual,0
297,brio,2015,4.00,5.9,60000,Petrol,Dealer,Manual,0
298,city,2009,3.35,11.0,87934,Petrol,Dealer,Manual,0
299,city,2017,11.50,12.5,9000,Diesel,Dealer,Manual,0
300,brio,2016,5.30,5.9,5464,Petrol,Dealer,Manual,0


In [6]:
df.describe()

,Year,Selling_Price,Present_Price,Kms_Driven,Owner
count,301.000000,301.000000,301.000000,301.000000,301.000000
mean,2013.627907,4.661296,7.628472,36947.205980,0.043189
std,2.891554,5.082812,8.644115,38886.883882,0.247915
min,2003.000000,0.100000,0.320000,500.000000,0.000000
25%,2012.000000,0.900000,1.200000,15000.000000,0.000000
50%,2014.000000,3.600000,6.400000,32000.000000,0.000000
75%,2016.000000,6.000000,9.900000,48767.000000,0.000000
max,2018.000000,35.000000,92.600000,500000.000000,3.000000


In [7]:
df.isnull().sum()

,0
Car_Name,0
Year,0
Selling_Price,0
Present_Price,0
Kms_Driven,0
Fuel_Type,0
Seller_Type,0
Transmission,0
Owner,0


In [8]:
df.duplicated().sum()

np.int64(2)

In [9]:
df[df.duplicated(keep=False)]

,Car_Name,Year,Selling_Price,Present_Price,Kms_Driven,Fuel_Type,Seller_Type,Transmission,Owner
15,ertiga,2016,7.75,10.79,43000,Diesel,Dealer,Manual,0
17,ertiga,2016,7.75,10.79,43000,Diesel,Dealer,Manual,0
51,fortuner,2015,23.00,30.61,40000,Diesel,Dealer,Automatic,0
93,fortuner,2015,23.00,30.61,40000,Diesel,Dealer,Automatic,0


In [10]:
print("Fuel Type:")
print(df["Fuel_Type"].unique())

print("\nSeller Type:")
print(df["Seller_Type"].unique())

print("\nTransmission:")
print(df["Transmission"].unique())

Fuel Type:
['Petrol' 'Diesel' 'CNG']

Seller Type:
['Dealer' 'Individual']

Transmission:
['Manual' 'Automatic']


In [11]:
df[['Selling_Price', 'Present_Price']].head(10)

,Selling_Price,Present_Price
0,3.35,5.59
1,4.75,9.54
2,7.25,9.85
3,2.85,4.15
4,4.60,6.87
5,9.25,9.83
6,6.75,8.12
7,6.50,8.61
8,8.75,8.89
9,7.45,8.92


In [12]:
numerical_features = df.select_dtypes(include=['int64', 'float64']).columns.tolist()
categorical_features = df.select_dtypes(include=['object']).columns.tolist()

print("Numerical Features:")
print(numerical_features)

print("\nCategorical Features:")
print(categorical_features)

Numerical Features:
['Year', 'Selling_Price', 'Present_Price', 'Kms_Driven', 'Owner']

Categorical Features:
['Car_Name', 'Fuel_Type', 'Seller_Type', 'Transmission']


## Data Dictionary – Cardekho Dataset

| Column Name | Meaning | Data Type | Role |
|---|---|---|---|
| `Car_Name` | Name of the vehicle | Object | Feature |
| `Year` | Manufacturing year of the vehicle | Integer | Feature |
| `Selling_Price` | Selling price of the used vehicle | Float | Target |
| `Present_Price` | Present/current price of the vehicle | Float | Feature |
| `Kms_Driven` | Kilometres driven by the vehicle | Integer | Feature |
| `Fuel_Type` | Fuel type of the vehicle | Object | Feature |
| `Seller_Type` | Type of seller | Object | Feature |
| `Transmission` | Transmission type of the vehicle | Object | Feature |
| `Owner` | Number of previous owners | Integer | Feature |

In [13]:
print("Fuel Type Counts:")
print(df["Fuel_Type"].value_counts())

print("\nSeller Type Counts:")
print(df["Seller_Type"].value_counts())

print("\nTransmission Counts:")
print(df["Transmission"].value_counts())

Fuel Type Counts:
Fuel_Type
Petrol    239
Diesel     60
CNG         2
Name: count, dtype: int64

Seller Type Counts:
Seller_Type
Dealer        195
Individual    106
Name: count, dtype: int64

Transmission Counts:
Transmission
Manual       261
Automatic     40
Name: count, dtype: int64


## Data Dictionary – Cardekho Dataset

| Column Name | Meaning | Data Type | Role |
|---|---|---|---|
| `Car_Name` | Name of the vehicle | Object | Feature |
| `Year` | Manufacturing year of the vehicle | Integer | Feature |
| `Selling_Price` | Selling price of the used vehicle | Float | Target |
| `Present_Price` | Present/current price of the vehicle | Float | Feature |
| `Kms_Driven` | Kilometres driven by the vehicle | Integer | Feature |
| `Fuel_Type` | Fuel type of the vehicle | Object | Feature |
| `Seller_Type` | Type of seller | Object | Feature |
| `Transmission` | Transmission type of the vehicle | Object | Feature |
| `Owner` | Number of previous owners | Integer | Feature |

## Initial Findings – Cardekho Dataset

- The dataset contains 301 vehicle records and 9 columns.
- The dataset contains both numerical and categorical features.
- The target variable for price prediction is `Selling_Price`.
- Numerical features include `Year`, `Selling_Price`, `Present_Price`, `Kms_Driven`, and `Owner`.
- Categorical features include `Car_Name`, `Fuel_Type`, `Seller_Type`, and `Transmission`.
- No missing values were identified in any of the 9 columns.
- Two duplicate records were identified in the original dataset.
- The dataset contains three fuel types: Petrol, Diesel, and CNG.
- The dataset contains two seller types: Dealer and Individual.
- The dataset contains two transmission types: Manual and Automatic.
- `Kms_Driven` ranges from 500 to 500,000 kilometres, indicating a wide variation in vehicle usage.
- `Selling_Price` ranges from 0.10 to 35.00, while `Present_Price` ranges from 0.32 to 92.60.

In [14]:
craigslist_df = pd.read_csv('/content/vehicles.csv')

craigslist_df.head()

ParserError: Error tokenizing data. C error: EOF inside string starting at row 2325

In [15]:
craigslist_sample = pd.read_csv(
    '/content/vehicles.csv',
    nrows=5
)

craigslist_sample.head()

,id,url,region,region_url,price,year,manufacturer,model,condition,cylinders,...,size,type,paint_color,image_url,description,county,state,lat,long,posting_date
0,7222695916,https://prescott.craigslist.org/cto/d/prescott...,prescott,https://prescott.craigslist.org,6000,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,az,NaN,NaN,NaN
1,7218891961,https://fayar.craigslist.org/ctd/d/bentonville...,fayetteville,https://fayar.craigslist.org,11900,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,ar,NaN,NaN,NaN
2,7221797935,https://keys.craigslist.org/cto/d/summerland-k...,florida keys,https://keys.craigslist.org,21000,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,fl,NaN,NaN,NaN
3,7222270760,https://worcester.craigslist.org/cto/d/west-br...,worcester / central MA,https://worcester.craigslist.org,1500,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,ma,NaN,NaN,NaN
4,7210384030,https://greensboro.craigslist.org/cto/d/trinit...,greensboro,https://greensboro.craigslist.org,4900,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,nc,NaN,NaN,NaN


In [16]:
craigslist_sample.columns.tolist()

['id',
 'url',
 'region',
 'region_url',
 'price',
 'year',
 'manufacturer',
 'model',
 'condition',
 'cylinders',
 'fuel',
 'odometer',
 'title_status',
 'transmission',
 'VIN',
 'drive',
 'size',
 'type',
 'paint_color',
 'image_url',
 'description',
 'county',
 'state',
 'lat',
 'long',
 'posting_date']

In [17]:
craigslist_sample[['price', 'manufacturer', 'model', 'condition', 'description']]

,price,manufacturer,model,condition,description
0,6000,NaN,NaN,NaN,NaN
1,11900,NaN,NaN,NaN,NaN
2,21000,NaN,NaN,NaN,NaN
3,1500,NaN,NaN,NaN,NaN
4,4900,NaN,NaN,NaN,NaN


In [19]:
craigslist_sample = pd.read_csv('/content/vehicles.csv', nrows=1000)

print("Missing descriptions:", craigslist_sample['description'].isnull().sum())
print("Total records checked:", len(craigslist_sample))

Missing descriptions: 27
Total records checked: 1000


In [20]:
print("Dataset Shape:", craigslist_sample.shape)

Dataset Shape: (1000, 26)


In [21]:
relevant_columns = [
    'price',
    'year',
    'manufacturer',
    'model',
    'condition',
    'cylinders',
    'fuel',
    'odometer',
    'transmission',
    'drive',
    'type',
    'description',
    'state',
    'posting_date'
]

craigslist_sample[relevant_columns].head()

,price,year,manufacturer,model,condition,cylinders,fuel,odometer,transmission,drive,type,description,state,posting_date
0,6000,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,az,NaN
1,11900,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,ar,NaN
2,21000,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,fl,NaN
3,1500,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,ma,NaN
4,4900,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,nc,NaN


In [22]:
craigslist_sample[relevant_columns].info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 14 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   price         1000 non-null   int64  
 1   year          963 non-null    float64
 2   manufacturer  903 non-null    object 
 3   model         965 non-null    object 
 4   condition     580 non-null    object 
 5   cylinders     653 non-null    object 
 6   fuel          960 non-null    object 
 7   odometer      970 non-null    float64
 8   transmission  973 non-null    object 
 9   drive         687 non-null    object 
 10  type          707 non-null    object 
 11  description   973 non-null    object 
 12  state         1000 non-null   object 
 13  posting_date  973 non-null    object 
dtypes: float64(2), int64(1), object(11)
memory usage: 109.5+ KB


In [23]:
missing_summary = craigslist_sample[relevant_columns].isnull().sum()

missing_summary = missing_summary[missing_summary > 0]

missing_summary

,0
year,37
manufacturer,97
model,35
condition,420
cylinders,347
fuel,40
odometer,30
transmission,27
drive,313
type,293


In [24]:
print("Missing descriptions:", craigslist_sample['description'].isnull().sum())
print("Total records checked:", len(craigslist_sample))

Missing descriptions: 27
Total records checked: 1000


## Data Dictionary – Craigslist Vehicles Dataset

| Column Name | Meaning | Data Type | Role |
|---|---|---|---|
| price | Listed price of the vehicle | Integer | Feature |
| year | Manufacturing year of the vehicle | Float | Feature |
| manufacturer | Vehicle manufacturer/brand | Object | Feature |
| model | Vehicle model | Object | Feature |
| condition | Condition of the vehicle | Object | Feature |
| cylinders | Number/type of engine cylinders | Object | Feature |
| fuel | Fuel type of the vehicle | Object | Feature |
| odometer | Vehicle mileage | Float | Feature |
| transmission | Transmission type | Object | Feature |
| drive | Drivetrain type | Object | Feature |
| type | Type/category of vehicle | Object | Feature |
| description | Textual description of the vehicle listing | Object | Text / NLP |
| state | State/location of the vehicle listing | Object | Feature |
| posting_date | Date/time when the listing was posted | Object | Feature |

## Initial Findings – Craigslist Vehicles Dataset

- The Craigslist dataset contains 21,972 vehicle records and 26 columns.
- The first 1,000 records were inspected for initial data understanding.
- The dataset contains both numerical and categorical vehicle-related attributes.
- The `description` column contains textual vehicle listing information and will be used for the later NLP/TF-IDF stage.
- Among the first 1,000 records, 27 records had missing vehicle descriptions.
- Missing values were observed in several vehicle attributes. The highest missing-value percentages were found in condition (42.0%), cylinders (34.7%), drive (31.3%), and type (29.3%).
- The `price` column was available for all 1,000 inspected records.
- The dataset contains vehicle information such as manufacturer, model, fuel type, mileage, transmission, drivetrain, vehicle type, and location.
- No data cleaning or modification was performed at this stage; the original dataset was retained for further processing.
- The most common vehicle types in the inspected sample were sedan, SUV, truck, and pickup.
- Among vehicles with available condition information, good-condition vehicles were the most common.

In [25]:
# Check total number of records in the Craigslist dataset
import csv

with open('/content/vehicles.csv', 'r', encoding='utf-8', errors='ignore') as f:
    reader = csv.reader(f)
    total_rows = sum(1 for row in reader) - 1

print("Total records in Craigslist dataset:", total_rows)

Total records in Craigslist dataset: 21972


In [27]:
# Check duplicate records in the Craigslist dataset

print("Duplicate records:", craigslist_sample.duplicated().sum())

Duplicate records: 0


In [28]:
# Check missing values in the Craigslist dataset

print("Missing values in each column:")
print(craigslist_sample[relevant_columns].isnull().sum())

Missing values in each column:
price             0
year             37
manufacturer     97
model            35
condition       420
cylinders       347
fuel             40
odometer         30
transmission     27
drive           313
type            293
description      27
state             0
posting_date     27
dtype: int64


In [29]:
# Calculate missing value percentages

missing_percentage = (
    craigslist_sample[relevant_columns].isnull().sum()
    / len(craigslist_sample) * 100
)

print("Missing value percentages:")
print(missing_percentage.round(2))

Missing value percentages:
price            0.0
year             3.7
manufacturer     9.7
model            3.5
condition       42.0
cylinders       34.7
fuel             4.0
odometer         3.0
transmission     2.7
drive           31.3
type            29.3
description      2.7
state            0.0
posting_date     2.7
dtype: float64


In [30]:
print("Highest missing-value columns:")
print(missing_percentage.sort_values(ascending=False).head(5))

Highest missing-value columns:
condition       42.0
cylinders       34.7
drive           31.3
type            29.3
manufacturer     9.7
dtype: float64


In [31]:
# Summary statistics for numerical columns

print("Summary statistics:")
print(craigslist_sample[relevant_columns].describe())

Summary statistics:
              price         year      odometer
count  1.000000e+03   963.000000  9.700000e+02
mean   1.005993e+06  2011.064382  1.164209e+05
std    3.123179e+07     8.959832  3.418087e+05
min    0.000000e+00  1929.000000  0.000000e+00
25%    5.940000e+03  2008.000000  3.132050e+04
50%    1.696900e+04  2013.000000  8.918750e+04
75%    2.866750e+04  2017.000000  1.594192e+05
max    9.876543e+08  2021.000000  9.999999e+06


In [32]:
# Check the range of numerical variables

print("Numerical variable ranges:")
print(craigslist_sample[["price", "year", "odometer"]].agg(["min", "max"]))

Numerical variable ranges:
         price    year   odometer
min          0  1929.0        0.0
max  987654321  2021.0  9999999.0


In [33]:
# Check categorical variable distributions

categorical_columns = [
    "manufacturer",
    "fuel",
    "transmission",
    "drive",
    "type",
    "condition"
]

for column in categorical_columns:
    print(f"\n{column} value counts:")
    print(craigslist_sample[column].value_counts().head(10))


manufacturer value counts:
manufacturer
ford             176
chevrolet        130
toyota            74
nissan            60
jeep              52
gmc               43
ram               33
honda             30
kia               26
mercedes-benz     25
Name: count, dtype: int64

fuel value counts:
fuel
gas       789
diesel     98
other      67
hybrid      6
Name: count, dtype: int64

transmission value counts:
transmission
automatic    702
other        221
manual        50
Name: count, dtype: int64

drive value counts:
drive
fwd    239
rwd    228
4wd    220
Name: count, dtype: int64

type value counts:
type
sedan          182
SUV            154
truck          109
pickup         105
other           37
coupe           27
hatchback       27
convertible     18
van             18
bus             14
Name: count, dtype: int64

condition value counts:
condition
good         380
excellent    121
like new      53
fair          25
new            1
Name: count, dtype: int64
